File: honjo-masamune/honjo/src/ast.ts
// Honjo Masamune — Abstract Syntax Tree
// Mirrors the EBNF grammar of the spec (§4 "Grammar").

export interface Pos { line: number; col: number; }

// ---- Expressions ----
export type Expr =
  | NumLit
  | StrLit
  | Ref
  | CutExpr
  | BondExpr
  | CloseExpr
  | CallExpr;

export interface NumLit { tag: "num"; value: number; floor?: number; pos: Pos; }
export interface StrLit { tag: "str"; value: string; pos: Pos; }
export interface Ref { tag: "ref"; name: string; pos: Pos; }

// cut <number|ident>            — arity-1 cut (generation)
export interface CutExpr { tag: "cut"; arg: Expr; pos: Pos; }

// a ~ b  [when cond]            — arity-2 cut (bond)
export interface BondExpr { tag: "bond"; left: Expr; right: Expr; guard?: Cond; pos: Pos; }

// close X(a1,...,ak) [by ident] — cut-to-closure (compound)
export interface CloseExpr { tag: "close"; central: string; args: Arg[]; by?: string; pos: Pos; }

// qname(args)                   — stdlib / module call
export interface CallExpr { tag: "call"; name: string[]; args: Arg[]; pos: Pos; }

export interface Arg { label?: string; value: Expr; }

export interface Cond { tag: "cond"; left: Expr; op: RelOp; right: Expr; pos: Pos; }
export type RelOp = ">" | "<" | ">=" | "<=" | "==";

// ---- Statements / declarations ----
export type Decl = FloorDecl | ImportDecl | ModuleDecl | Stmt;
export type Stmt = BindStmt | TrackStmt | ObserveStmt | AssertStmt | ExprStmt;

export interface FloorDecl { tag: "floor"; value: number; pos: Pos; }
export interface ImportDecl { tag: "import"; name: string[]; pos: Pos; }
export interface ModuleDecl { tag: "module"; name: string; body: Decl[]; pos: Pos; }

export interface BindStmt { tag: "bind"; name: string; value: Expr; pos: Pos; }
export interface ExprStmt { tag: "exprStmt"; expr: Expr; pos: Pos; }
export interface ObserveStmt { tag: "observe"; expr: Expr; as?: string; pos: Pos; }
export interface AssertStmt { tag: "assert"; cond: Cond; emit?: string; pos: Pos; }

// track x in P [with reps r1,...] until <admit> yield r
export interface TrackStmt {
  tag: "track";
  item: string;
  process: Expr;
  reps?: string[];
  admit: Admit;
  yieldName: string;
  pos: Pos;
}
export type Admit = { kind: "converge" } | { kind: "diverge" } | { kind: "cond"; cond: Cond };

export interface Program { decls: Decl[]; }

File: honjo-masamune/honjo-rs/src/ast.rs
//! Honjo Masamune — AST (mirrors EBNF grammar §4).

#[derive(Debug, Clone, Copy)]
pub struct Pos {
    pub line: usize,
    pub col: usize,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RelOp {
    Gt,
    Lt,
    Ge,
    Le,
    Eq,
}

#[derive(Debug, Clone)]
pub struct Cond {
    pub left: Expr,
    pub op: RelOp,
    pub right: Expr,
    pub pos: Pos,
}

#[derive(Debug, Clone)]
pub enum Admit {
    Converge,
    Diverge,
    Cond(Box<Cond>),
}

#[derive(Debug, Clone)]
pub struct Arg {
    pub label: Option<String>,
    pub value: Expr,
}

#[derive(Debug, Clone)]
pub enum Expr {
    Num { value: f64, floor: Option<f64>, pos: Pos },
    Str { value: String, pos: Pos },
    Ref { name: String, pos: Pos },
    Cut { arg: Box<Expr>, pos: Pos },
    Bond { left: Box<Expr>, right: Box<Expr>, guard: Option<Box<Cond>>, pos: Pos },
    Close { central: String, args: Vec<Arg>, by: Option<String>, pos: Pos },
    Call { name: Vec<String>, args: Vec<Arg>, pos: Pos },
    TrackExpr {
        item: String,
        process: Box<Expr>,
        reps: Option<Vec<String>>,
        admit: Admit,
        role: Option<String>,
        pos: Pos,
    },
}

impl Expr {
    pub fn pos(&self) -> Pos {
        match self {
            Expr::Num { pos, .. }
            | Expr::Str { pos, .. }
            | Expr::Ref { pos, .. }
            | Expr::Cut { pos, .. }
            | Expr::Bond { pos, .. }
            | Expr::Close { pos, .. }
            | Expr::Call { pos, .. }
            | Expr::TrackExpr { pos, .. } => *pos,
        }
    }
}

#[derive(Debug, Clone)]
pub enum Stmt {
    Bind { name: String, value: Expr, pos: Pos },
    ExprStmt { expr: Expr, pos: Pos },
    Observe { expr: Expr, as_name: Option<String>, pos: Pos },
    Assert { cond: Cond, emit: Option<String>, pos: Pos },
    Track {
        item: String,
        process: Expr,
        reps: Option<Vec<String>>,
        admit: Admit,
        yield_name: String,
        pos: Pos,
    },
}

#[derive(Debug, Clone)]
pub enum Decl {
    Floor { value: f64, pos: Pos },
    Import { name: Vec<String>, pos: Pos },
    Module { name: String, body: Vec<Decl>, pos: Pos },
    Stmt(Stmt),
}

#[derive(Debug, Clone)]
pub struct Program {
    pub decls: Vec<Decl>,
}
